File: ios/LingoPlay/SourceSeparationModel.swift
import Foundation
@preconcurrency import SWCompression

enum SourceSeparationManifest {
    static let version = "spleeter-2stems-fp16-c6c5c430"
    static let archiveRoot = "sherpa-onnx-spleeter-2stems-fp16"
    static let archiveName = archiveRoot + ".tar.bz2"
    static let archiveBytes: Int64 = 35_271_738
    static let archiveSHA256 = "c6c5c4307673bc6813ddf58d4efdff57c26d2dfc3f25b05c7a32db453d70aca6"
    static let vocalsName = "vocals.fp16.onnx"
    static let accompanimentName = "accompaniment.fp16.onnx"
    static let archiveURL = "https://github.com/k2-fsa/sherpa-onnx/releases/download/source-separation-models/\(archiveName)"
    static let archiveSpec = PinnedDownloadSpec(
        name: archiveName, url: archiveURL, bytes: archiveBytes, sha256: archiveSHA256
    )
}

enum SourceSeparationArchivePolicy {
    static let maximumEntries = 16
    static let maximumUncompressedBytes = 96 * 1024 * 1024
    static let maximumEntryBytes = 48 * 1024 * 1024
    private static let allowed = Set([SourceSeparationManifest.vocalsName, SourceSeparationManifest.accompanimentName])

    static func relativePath(for entryName: String) -> String? {
        guard !entryName.isEmpty, entryName.count <= 256,
              !entryName.contains("\0"), !entryName.contains("\\")
        else { return nil }
        if entryName == SourceSeparationManifest.archiveRoot { return "" }
        let prefix = SourceSeparationManifest.archiveRoot + "/"
        guard entryName.hasPrefix(prefix) else { return nil }
        let relative = String(entryName.dropFirst(prefix.count)).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !relative.isEmpty else { return "" }
        guard !relative.contains("/"), relative != ".", relative != "..", allowed.contains(relative) else { return nil }
        return relative
    }
}

struct InstalledSourceSeparationModel: Sendable, Equatable {
    let vocalsURL: URL
    let accompanimentURL: URL
}

struct SourceSeparationModelStore {
    private static let activePointerName = "active-model.txt"
    private let fileManager = FileManager.default

    func model() -> InstalledSourceSeparationModel? {
        guard let root = try? rootURL(create: false),
              let version = try? String(contentsOf: root.appendingPathComponent(Self.activePointerName), encoding: .utf8)
                .trimmingCharacters(in: .whitespacesAndNewlines),
              version == SourceSeparationManifest.version
        else { return nil }
        return validatedModel(at: root.appendingPathComponent(version, isDirectory: true))
    }

    func validatedModel(at directory: URL) -> InstalledSourceSeparationModel? {
        let marker = directory.appendingPathComponent("pack.sha256")
        guard (try? String(contentsOf: marker, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines)) == SourceSeparationManifest.archiveSHA256
        else { return nil }
        let vocals = directory.appendingPathComponent(SourceSeparationManifest.vocalsName)
        let accompaniment = directory.appendingPathComponent(SourceSeparationManifest.accompanimentName)
        guard isNonEmptyFile(vocals), isNonEmptyFile(accompaniment) else { return nil }
        return InstalledSourceSeparationModel(vocalsURL: vocals, accompanimentURL: accompaniment)
    }

    func rootURL(create: Bool) throws -> URL {
        guard let support = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else {
            throw PinnedDownloadError.activationFailed
        }
        let root = support.appendingPathComponent("LingoPlay/Models/SourceSeparation", isDirectory: true)
        if create { try fileManager.createDirectory(at: root, withIntermediateDirectories: true) }
        return root
    }

    private func isNonEmptyFile(_ url: URL) -> Bool {
        guard let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]) else { return false }
        return values.isRegularFile == true && (values.fileSize ?? 0) > 0
    }
}

actor SourceSeparationModelInstaller {
    private static let activePointerName = "active-model.txt"
    private static let storageSafetyMarginBytes: Int64 = 64 * 1024 * 1024
    private let fileManager = FileManager.default
    private let store = SourceSeparationModelStore()

    func state() -> ASRModelInstallState {
        guard let model = store.model() else { return .notInstalled }
        let bytes = (PinnedFileIntegrity.fileSize(model.vocalsURL) ?? 0) +
            (PinnedFileIntegrity.fileSize(model.accompanimentURL) ?? 0)
        return .installed(bytes: bytes)
    }

    func install(
        wifiOnly: Bool,
        progress: @escaping @Sendable (Double) async -> Void
    ) async throws -> InstalledSourceSeparationModel {
        if wifiOnly {
            guard await ModelNetworkPolicy.isUsingWiFi() else {
                throw NSError(
                    domain: "LingoPlay.SourceSeparation", code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Connect to Wi-Fi or disable ‘Download models on Wi-Fi only’ before installing Clean Background."]
                )
            }
        }
        if let installed = store.model() {
            await progress(1)
            return installed
        }

        let root = try store.rootURL(create: true)
        try ensureStorage(at: root)
        let archive = try await PinnedModelDownload.downloadVerified(
            spec: SourceSeparationManifest.archiveSpec, root: root
        ) { bytes in
            await progress(Double(bytes) / Double(SourceSeparationManifest.archiveBytes))
        }
        try Task.checkCancellation()

        let staging = root.appendingPathComponent(SourceSeparationManifest.version + ".staging", isDirectory: true)
        let versionDirectory = root.appendingPathComponent(SourceSeparationManifest.version, isDirectory: true)
        try? fileManager.removeItem(at: staging)
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: true)
        do {
            try extractVerifiedArchive(archive, to: staging)
            try SourceSeparationManifest.archiveSHA256.write(
                to: staging.appendingPathComponent("pack.sha256"), atomically: true, encoding: .utf8
            )
            guard store.validatedModel(at: staging) != nil else { throw PinnedDownloadError.integrityFailed }
            try? fileManager.removeItem(at: versionDirectory)
            try fileManager.moveItem(at: staging, to: versionDirectory)
            try writeActivePointer(root: root)
            try? fileManager.removeItem(at: archive)
            guard let activated = store.model() else { throw PinnedDownloadError.activationFailed }
            await progress(1)
            return activated
        } catch {
            try? fileManager.removeItem(at: staging)
            throw error
        }
    }

    func deleteInstalledModel() throws {
        let root = try store.rootURL(create: false)
        if fileManager.fileExists(atPath: root.path) { try fileManager.removeItem(at: root) }
    }

    private func extractVerifiedArchive(_ archive: URL, to staging: URL) throws {
        try Task.checkCancellation()
        let compressed = try Data(contentsOf: archive, options: [.mappedIfSafe])
        guard compressed.count == Int(SourceSeparationManifest.archiveBytes) else { throw PinnedDownloadError.integrityFailed }
        let tarData = try BZip2.decompress(data: compressed)
        try Task.checkCancellation()
        guard tarData.count <= SourceSeparationArchivePolicy.maximumUncompressedBytes else { throw PinnedDownloadError.integrityFailed }
        let entries = try TarContainer.open(container: tarData)
        guard entries.count <= SourceSeparationArchivePolicy.maximumEntries else { throw PinnedDownloadError.integrityFailed }
        let stagingPrefix = staging.standardizedFileURL.path + "/"
        var extractedBytes = 0
        for entry in entries {
            try Task.checkCancellation()
            let isDirectory = entry.info.type == .directory
            let isRegular = entry.info.type == .regular || entry.info.type == .contiguous
            guard isDirectory || isRegular,
                  let relative = SourceSeparationArchivePolicy.relativePath(for: entry.info.name)
            else { throw PinnedDownloadError.integrityFailed }
            guard !relative.isEmpty else { continue }
            let size = entry.data?.count ?? 0
            guard size <= SourceSeparationArchivePolicy.maximumEntryBytes else { throw PinnedDownloadError.integrityFailed }
            let output = staging.appendingPathComponent(relative).standardizedFileURL
            guard output.path.hasPrefix(stagingPrefix) else { throw PinnedDownloadError.integrityFailed }
            if isDirectory { continue }
            guard let data = entry.data, entry.info.size == data.count else { throw PinnedDownloadError.integrityFailed }
            extractedBytes += data.count
            guard extractedBytes <= SourceSeparationArchivePolicy.maximumUncompressedBytes else { throw PinnedDownloadError.integrityFailed }
            try data.write(to: output, options: [.atomic])
        }
    }

    private func ensureStorage(at root: URL) throws {
        let values = try root.resourceValues(forKeys: [.volumeAvailableCapacityForImportantUsageKey])
        let required = SourceSeparationManifest.archiveBytes +
            Int64(SourceSeparationArchivePolicy.maximumUncompressedBytes) + Self.storageSafetyMarginBytes
        if let available = values.volumeAvailableCapacityForImportantUsage, available < required {
            throw NSError(
                domain: "LingoPlay.SourceSeparation", code: 2,
                userInfo: [NSLocalizedDescriptionKey: "Not enough free storage to install Clean Background."]
            )
        }
    }

    private func writeActivePointer(root: URL) throws {
        let pointer = root.appendingPathComponent(Self.activePointerName)
        let temporary = root.appendingPathComponent(Self.activePointerName + ".tmp")
        try SourceSeparationManifest.version.write(to: temporary, atomically: true, encoding: .utf8)
        if fileManager.fileExists(atPath: pointer.path) {
            _ = try fileManager.replaceItemAt(pointer, withItemAt: temporary)
        } else {
            try fileManager.moveItem(at: temporary, to: pointer)
        }
    }
}

File: android/app/src/main/java/com/lingoplay/app/SourceSeparationModel.kt
package com.lingoplay.app

import android.content.Context
import android.os.StatFs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.compress.compressors.bzip2.BZip2CompressorInputStream
import java.io.File
import java.io.FileInputStream
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

internal object SourceSeparationManifest {
    const val version = "spleeter-2stems-fp16-c6c5c430"
    const val archiveRoot = "sherpa-onnx-spleeter-2stems-fp16"
    const val archiveName = "$archiveRoot.tar.bz2"
    const val archiveBytes = 35_271_738L
    const val archiveSha256 = "d54561979bd2e08a51e7dbd99ac36bb47564e089eefd403636dbca93e811bba2"
    const val previousArchiveSha256 = "c6c5c4307673bc6813ddf58d4efdff57c26d2dfc3f25b05c7a32db453d70aca6"
    const val vocalsName = "vocals.fp16.onnx"
    const val vocalsBytes = 19_681_017L
    const val vocalsSha256 = "24cef84aedcd1fe87c0b743ef3370ad34dc1fabf6c9014d6128a75a538c7b668"
    const val accompanimentName = "accompaniment.fp16.onnx"
    const val accompanimentBytes = 19_681_024L
    const val accompanimentSha256 = "d14cea55793cc531a5875f5f4da08207d1c5ab9292e8e0099a104eecb014fcc0"
    const val archiveUrl =
        "https://github.com/k2-fsa/sherpa-onnx/releases/download/source-separation-models/$archiveName"

    val archiveSpec = ModelFileSpec(archiveName, archiveUrl, archiveBytes, archiveSha256)
    val acceptedArchiveSha256 = setOf(archiveSha256, previousArchiveSha256)
    val vocalsSpec = ModelFileSpec(vocalsName, "", vocalsBytes, vocalsSha256)
    val accompanimentSpec = ModelFileSpec(accompanimentName, "", accompanimentBytes, accompanimentSha256)

    fun archiveMatches(file: File): Boolean =
        file.isFile && file.length() == archiveBytes && ModelIntegrity.sha256(file) in acceptedArchiveSha256
}

internal object SourceSeparationArchivePolicy {
    const val maxEntries = 16
    const val maxUncompressedBytes = 96L * 1024L * 1024L
    const val maxEntryBytes = 48L * 1024L * 1024L
    private val allowedFiles = setOf(SourceSeparationManifest.vocalsName, SourceSeparationManifest.accompanimentName)

    fun relativePath(entryName: String): String? {
        if (entryName.isEmpty() || entryName.length > 256 || '\u0000' in entryName || '\\' in entryName) return null
        if (entryName == SourceSeparationManifest.archiveRoot) return ""
        val prefix = "${SourceSeparationManifest.archiveRoot}/"
        if (!entryName.startsWith(prefix)) return null
        val relative = entryName.removePrefix(prefix).trimEnd('/')
        if (relative.isEmpty()) return ""
        if ('/' in relative || relative == "." || relative == "..") return null
        return relative.takeIf { it in allowedFiles }
    }

    fun allowsEntry(entryName: String, isDirectory: Boolean, isRegularFile: Boolean, size: Long): Boolean {
        if (!isDirectory && !isRegularFile) return false
        if (size < 0 || size > maxEntryBytes) return false
        return relativePath(entryName) != null
    }
}

data class SpleeterSourceSeparationModel(
    val vocals: File,
    val accompaniment: File,
)

object SourceSeparationModelStore {
    fun find(context: Context): SpleeterSourceSeparationModel? =
        SourceSeparationModelInstaller.activeDirectory(context)?.let(::validatedModel)

    internal fun validatedModel(root: File): SpleeterSourceSeparationModel? {
        val marker = File(root, "pack.sha256")
        if (marker.takeIf(File::isFile)?.readText()?.trim() !in SourceSeparationManifest.acceptedArchiveSha256) return null
        val vocals = File(root, SourceSeparationManifest.vocalsName)
        val accompaniment = File(root, SourceSeparationManifest.accompanimentName)
        if (!vocals.isFile || vocals.length() != SourceSeparationManifest.vocalsBytes ||
            !accompaniment.isFile || accompaniment.length() != SourceSeparationManifest.accompanimentBytes
        ) return null
        return SpleeterSourceSeparationModel(vocals, accompaniment)
    }

    internal fun contentHashesMatch(root: File): Boolean =
        ModelIntegrity.matches(File(root, SourceSeparationManifest.vocalsName), SourceSeparationManifest.vocalsSpec) &&
            ModelIntegrity.matches(
                File(root, SourceSeparationManifest.accompanimentName),
                SourceSeparationManifest.accompanimentSpec,
            )
}

object SourceSeparationModelInstaller {
    private const val rootPath = "lingoplay/models/source-separation"
    private const val activePointerName = "active-model.txt"
    private const val storageSafetyMarginBytes = 64L * 1024L * 1024L

    fun state(context: Context): ModelInstallState {
        val model = SourceSeparationModelStore.find(context) ?: return ModelInstallState.NotInstalled
        val bytes = model.vocals.length() + model.accompaniment.length()
        return ModelInstallState.Installed(bytes)
    }

    suspend fun install(
        context: Context,
        wifiOnly: Boolean,
        onProgress: suspend (ModelInstallState.Downloading) -> Unit,
    ): SpleeterSourceSeparationModel = withContext(Dispatchers.IO) {
        if (wifiOnly && !ASRModelInstaller.isOnWifi(context)) {
            throw IllegalStateException(
                "Connect to Wi-Fi or disable ‘Download models on Wi-Fi only’ before installing Clean Background.",
            )
        }
        SourceSeparationModelStore.find(context)?.let {
            onProgress(ModelInstallState.Downloading(SourceSeparationManifest.archiveBytes, SourceSeparationManifest.archiveBytes))
            return@withContext it
        }

        val root = File(context.filesDir, rootPath).apply { mkdirs() }
        val archive = File(root, SourceSeparationManifest.archiveName)
        val part = File(root, "${SourceSeparationManifest.archiveName}.part")
        if (archive.exists() && !SourceSeparationManifest.archiveMatches(archive)) archive.delete()
        if (part.length() > SourceSeparationManifest.archiveBytes ||
            (part.length() == SourceSeparationManifest.archiveBytes && !SourceSeparationManifest.archiveMatches(part))
        ) part.delete()

        val verified = SourceSeparationManifest.archiveMatches(archive)
        val current = if (verified) SourceSeparationManifest.archiveBytes else part.length().coerceAtLeast(0L)
        ensureStorage(context, SourceSeparationManifest.archiveBytes - current)
        onProgress(ModelInstallState.Downloading(current, SourceSeparationManifest.archiveBytes))

        if (!verified) {
            ASRModelInstaller.downloadResumable(SourceSeparationManifest.archiveSpec, part) { bytes ->
                currentCoroutineContext().ensureActive()
                onProgress(ModelInstallState.Downloading(bytes, SourceSeparationManifest.archiveBytes))
            }
            if (!SourceSeparationManifest.archiveMatches(part)) {
                val actualBytes = part.length()
                val actualSha256 = part.takeIf(File::isFile)?.let(ModelIntegrity::sha256).orEmpty()
                part.delete()
                error(
                    "Clean Background model failed integrity verification " +
                        "(bytes=$actualBytes, sha256=$actualSha256).",
                )
            }
            if (archive.exists() && !archive.delete()) error("Unable to replace Clean Background archive.")
            if (!part.renameTo(archive)) error("Unable to prepare Clean Background archive.")
        }

        val staging = File(root, "${SourceSeparationManifest.version}.staging")
        val versionDir = File(root, SourceSeparationManifest.version)
        if (staging.exists() && !staging.deleteRecursively()) error("Unable to clear Clean Background staging directory.")
        if (!staging.mkdirs() && !staging.isDirectory) error("Unable to create Clean Background staging directory.")
        try {
            extractVerifiedArchive(archive, staging)
            if (!SourceSeparationModelStore.contentHashesMatch(staging)) {
                error("Clean Background extracted models failed exact size/SHA-256 verification.")
            }
            File(staging, "pack.sha256").writeText(ModelIntegrity.sha256(archive))
            SourceSeparationModelStore.validatedModel(staging)
                ?: error("The verified Clean Background archive is incomplete.")
            if (versionDir.exists() && !versionDir.deleteRecursively()) error("Unable to replace Clean Background model.")
            if (!staging.renameTo(versionDir)) error("Unable to activate Clean Background model.")
            writeActivePointer(root, SourceSeparationManifest.version)
            archive.delete()
            val activated = SourceSeparationModelStore.find(context)
                ?: error("Verified Clean Background model could not be activated.")
            onProgress(ModelInstallState.Downloading(SourceSeparationManifest.archiveBytes, SourceSeparationManifest.archiveBytes))
            activated
        } catch (error: Throwable) {
            staging.deleteRecursively()
            throw error
        }
    }

    fun deleteInstalled(context: Context) {
        val root = File(context.filesDir, rootPath)
        if (root.exists() && !root.deleteRecursively()) error("Unable to delete the installed Clean Background model.")
    }

    internal fun activeDirectory(context: Context): File? {
        val root = File(context.filesDir, rootPath)
        val pointer = File(root, activePointerName)
        val version = pointer.takeIf(File::isFile)?.readText()?.trim().orEmpty()
        if (version != SourceSeparationManifest.version) return null
        return File(root, version).takeIf(File::isDirectory)
    }

    private suspend fun extractVerifiedArchive(archive: File, staging: File) {
        currentCoroutineContext().ensureActive()
        var entries = 0
        var extractedBytes = 0L
        val stagingPrefix = staging.canonicalPath + File.separator
        TarArchiveInputStream(BZip2CompressorInputStream(FileInputStream(archive).buffered(), true)).use { tar ->
            while (true) {
                currentCoroutineContext().ensureActive()
                val entry = tar.nextEntry ?: break
                entries++
                if (entries > SourceSeparationArchivePolicy.maxEntries ||
                    !SourceSeparationArchivePolicy.allowsEntry(entry.name, entry.isDirectory, entry.isFile, entry.size)
                ) error("Clean Background archive contains an unsafe entry.")
                val relative = SourceSeparationArchivePolicy.relativePath(entry.name).orEmpty()
                if (relative.isEmpty()) continue
                val output = File(staging, relative)
                if (!output.canonicalPath.startsWith(stagingPrefix)) error("Clean Background archive escapes staging.")
                if (entry.isDirectory) continue
                extractedBytes += entry.size
                if (extractedBytes > SourceSeparationArchivePolicy.maxUncompressedBytes) {
                    error("Clean Background archive exceeds its extracted-size limit.")
                }
                output.outputStream().buffered().use { sink -> tar.copyTo(sink, 256 * 1024) }
                if (output.length() != entry.size) error("Clean Background archive entry was truncated.")
            }
        }
    }

    private fun ensureStorage(context: Context, remainingDownloadBytes: Long) {
        val required = remainingDownloadBytes.coerceAtLeast(0L) +
            SourceSeparationArchivePolicy.maxUncompressedBytes + storageSafetyMarginBytes
        if (StatFs(context.filesDir.absolutePath).availableBytes < required) {
            error("Not enough free storage to install Clean Background. Free at least ${MediaFormatting.bytes(required)}.")
        }
    }

    private fun writeActivePointer(root: File, version: String) {
        val pointer = File(root, activePointerName)
        val temporary = File(root, "$activePointerName.tmp")
        temporary.writeText(version)
        try {
            Files.move(temporary.toPath(), pointer.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        } catch (_: AtomicMoveNotSupportedException) {
            Files.move(temporary.toPath(), pointer.toPath(), StandardCopyOption.REPLACE_EXISTING)
        }
    }
}
